swap/txqueue/backend.go

```diff
@@ -0,0 +1,86 @@
+package txqueue
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind/backends"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+var (
+	// ErrTransactionReverted is given when the transaction that cashes a cheque is reverted
+	ErrTransactionReverted = errors.New("Transaction reverted")
+)
+
+// Backend is the minimum amount of functionality required by the underlying ethereum backend
+type Backend interface {
+	bind.ContractBackend
+	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+	TransactionByHash(ctx context.Context, txHash common.Hash) (*types.Transaction, bool, error)
+}
+
+// TestBackend is the backend to use for tests with a simulated backend
+type TestBackend struct {
+	*backends.SimulatedBackend
+}
+
+// WaitMined waits until either the transaction with the given hash has been mined or the context is cancelled
+func WaitMined(ctx context.Context, b Backend, hash common.Hash) (*types.Receipt, error) {
+	queryTicker := time.NewTicker(time.Second)
+	defer queryTicker.Stop()
+
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	for {
+		receipt, err := b.TransactionReceipt(ctx, hash)
+		if receipt != nil {
+			// indicate whether the transaction did not revert
+			if receipt.Status != types.ReceiptStatusSuccessful {
+				return nil, ErrTransactionReverted
+			}
+			return receipt, nil
+		}
+		if err != nil {
+			log.Trace("Receipt retrieval failed", "err", err)
+		} else {
+			log.Trace("Transaction not yet mined")
+		}
+		// Wait for the next round.
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-queryTicker.C:
+		}
+	}
+}
+
+// SendTransaction adds a commit after a successful send
+func (b *TestBackend) SendTransaction(ctx context.Context, tx *types.Transaction) (err error) {
+	err = b.SimulatedBackend.SendTransaction(ctx, tx)
+	if err == nil {
+		b.SimulatedBackend.Commit()
+	}
+	return err
+}
+
+// Close overrides Close so a SimulatedBackend can be used accross multiple tests for now
+func (b *TestBackend) Close() {
+
+}
```

Comment:
The function is not supposed to have an implementation. It overrides the `Close` function of the underlying `SimulatedBackend` so that it doesn't do anything. (and thereby the same `SimulatedBackend ` can be reused in the next test).

Reply:
Maybe add a log statement, so it is extra clear? I now also understand the comment.

Reply:
I made the comment a bit more explanatory.